## 1. 2178 미로탐색

#### 문제
N×M크기의 배열로 표현되는 미로가 있다.  
  
1	0	1	1	1	1  
1	0	1	0	1	0  
1	0	1	0	1	1  
1	1	1	0	1	1  
  
미로에서 1은 이동할 수 있는 칸을 나타내고, 0은 이동할 수 없는 칸을 나타낸다.   
이러한 미로가 주어졌을 때, (1, 1)에서 출발하여 (N, M)의 위치로 이동할 때 지나야 하는 최소의 칸 수를 구하는 프로그램을 작성하시오.  
한 칸에서 다른 칸으로 이동할 때, 서로 인접한 칸으로만 이동할 수 있다.  
  
위의 예에서는 15칸을 지나야 (N, M)의 위치로 이동할 수 있다. 칸을 셀 때에는 시작 위치와 도착 위치도 포함한다.  
  
#### 입력
첫째 줄에 두 정수 N, M(2 ≤ N, M ≤ 100)이 주어진다. 다음 N개의 줄에는 M개의 정수로 미로가 주어진다. 각각의 수들은 붙어서 입력으로 주어진다.  
  
#### 출력
첫째 줄에 지나야 하는 최소의 칸 수를 출력한다. 항상 도착위치로 이동할 수 있는 경우만 입력으로 주어진다.  

### ------------------------------------------------------------------------------

In [11]:
# import sys
# input = sys.stdin.readline
from collections import deque

dx = 0, 0, 1, -1
dy = 1, -1, 0, 0

N, M = map(int, input().split())
m = [input() for _ in range(N)]
v = [[False] * M for _ in range(N)]
l = 1
q = deque()
q.append((0, 0))
v[0][0] = True
print(f'map; {m}')
while q:
    for _ in range(len(q)):
        y, x = q.popleft()
        print(f'현재위치: {y}, {x}')
        print(f'현재이동거리: {l}')
        for d in range(4):
            cy = y + dy[d]
            cx = x + dx[d]
            print(f'검사: {cy}, {cx}')
            if cx>=0 and cx<M and cy>=0 and cy<N and m[cy][cx]=='1' and not v[cy][cx]:
                if (cy, cx) == (N-1, M-1):
                    print(l+1)
#                     exit()
                print(f'통과: {cy}, {cx}')
                q.append((cy,cx))
                v[cy][cx] = True            
    l += 1

4 6
101111
101010
101011
111011
map; ['101111', '101010', '101011', '111011']
현재위치: 0, 0
현재이동거리: 1
검사: 1, 0
통과: 1, 0
검사: -1, 0
검사: 0, 1
검사: 0, -1
현재위치: 1, 0
현재이동거리: 2
검사: 2, 0
통과: 2, 0
검사: 0, 0
검사: 1, 1
검사: 1, -1
현재위치: 2, 0
현재이동거리: 3
검사: 3, 0
통과: 3, 0
검사: 1, 0
검사: 2, 1
검사: 2, -1
현재위치: 3, 0
현재이동거리: 4
검사: 4, 0
검사: 2, 0
검사: 3, 1
통과: 3, 1
검사: 3, -1
현재위치: 3, 1
현재이동거리: 5
검사: 4, 1
검사: 2, 1
검사: 3, 2
통과: 3, 2
검사: 3, 0
현재위치: 3, 2
현재이동거리: 6
검사: 4, 2
검사: 2, 2
통과: 2, 2
검사: 3, 3
검사: 3, 1
현재위치: 2, 2
현재이동거리: 7
검사: 3, 2
검사: 1, 2
통과: 1, 2
검사: 2, 3
검사: 2, 1
현재위치: 1, 2
현재이동거리: 8
검사: 2, 2
검사: 0, 2
통과: 0, 2
검사: 1, 3
검사: 1, 1
현재위치: 0, 2
현재이동거리: 9
검사: 1, 2
검사: -1, 2
검사: 0, 3
통과: 0, 3
검사: 0, 1
현재위치: 0, 3
현재이동거리: 10
검사: 1, 3
검사: -1, 3
검사: 0, 4
통과: 0, 4
검사: 0, 2
현재위치: 0, 4
현재이동거리: 11
검사: 1, 4
통과: 1, 4
검사: -1, 4
검사: 0, 5
통과: 0, 5
검사: 0, 3
현재위치: 1, 4
현재이동거리: 12
검사: 2, 4
통과: 2, 4
검사: 0, 4
검사: 1, 5
검사: 1, 3
현재위치: 0, 5
현재이동거리: 12
검사: 1, 5
검사: -1, 5
검사: 0, 6
검사: 0, 4
현재위치: 2, 4
현재이동거리: 13
검사: 3, 4
통과: 3, 4
검사: 1, 4
검사

104ms.

4 6  
101111  
101010  
101011  
111011  

---
15

## 2. 7576 토마토

#### 문제
철수의 토마토 농장에서는 토마토를 보관하는 큰 창고를 가지고 있다.  
토마토는 아래의 그림과 같이 격자 모양 상자의 칸에 하나씩 넣어서 창고에 보관한다.   
  
<img src="https://www.acmicpc.net/upload/images/tmt.png">
  
창고에 보관되는 토마토들 중에는 잘 익은 것도 있지만, 아직 익지 않은 토마토들도 있을 수 있다.  
보관 후 하루가 지나면, 익은 토마토들의 인접한 곳에 있는 익지 않은 토마토들은 익은 토마토의 영향을 받아 익게 된다.  
하나의 토마토의 인접한 곳은 왼쪽, 오른쪽, 앞, 뒤 네 방향에 있는 토마토를 의미한다.  
대각선 방향에 있는 토마토들에게는 영향을 주지 못하며, 토마토가 혼자 저절로 익는 경우는 없다고 가정한다.  
철수는 창고에 보관된 토마토들이 며칠이 지나면 다 익게 되는지, 그 최소 일수를 알고 싶어 한다.  
  
토마토를 창고에 보관하는 격자모양의 상자들의 크기와 익은 토마토들과 익지 않은 토마토들의 정보가 주어졌을 때,  
며칠이 지나면 토마토들이 모두 익는지, 그 최소 일수를 구하는 프로그램을 작성하라. 단, 상자의 일부 칸에는 토마토가 들어있지 않을 수도 있다.  
  
#### 입력
첫 줄에는 상자의 크기를 나타내는 두 정수 M,N이 주어진다.  
M은 상자의 가로 칸의 수, N 은 상자의 세로 칸의 수를 나타낸다. 단, 2 ≤ M,N ≤ 1,000 이다.  
둘째 줄부터는 하나의 상자에 저장된 토마토들의 정보가 주어진다. 즉, 둘째 줄부터 N개의 줄에는 상자에 담긴 토마토의 정보가 주어진다.  
하나의 줄에는 상자 가로줄에 들어있는 토마토의 상태가 M개의 정수로 주어진다.  
정수 1은 익은 토마토, 정수 0은 익지 않은 토마토, 정수 -1은 토마토가 들어있지 않은 칸을 나타낸다.   
  
#### 출력
여러분은 토마토가 모두 익을 때까지의 최소 날짜를 출력해야 한다.  
만약, 저장될 때부터 모든 토마토가 익어있는 상태이면 0을 출력해야 하고, 토마토가 모두 익지는 못하는 상황이면 -1을 출력해야 한다.  

### ------------------------------------------------------------------------------

0이 없다면 모두 익어있는 상태. 바로 0 출력, 0근처가 모두 -1이면 -1출력.  
익어있는토마토 기준으로 탐색하지말고 안익어있는애들기준으로 탐색. 

In [18]:
# import sys
# input = sys.stdin.readline
from collections import deque
dx = 0, 0, 1, -1
dy = 1, -1, 0, 0
M, N = map(int, input().split())
m = []

# 안익은 애들 정보 저장하기..
unrp = set()
for j in range(N):
    r = list(map(int, input().split()))
    m.append(r)
    for i in range(M):
        if r[i]==0:
            unrp.add((j,i))
print(f'안익은 애들: {unrp}')
# 안익은 토마토가 없는경우 0 출력하고 끝            
if not unrp:
    print(0)
    exit()
    
days = 0                  # 며칠이 지났는지 저장..
changed = True           # 바뀌는 애들이 있으면.. 계속.

# 안익은 애들이 있고, 변화가 있는동안
while unrp and changed:
    days += 1              # 날짜 증가..
    changed = []
    ur_cp = unrp.copy()    # 반복돌 복사본.. 
    # 안익은 애들 전체를 돌면서...
    for cy, cx in ur_cp:
        impossible = True   # 근처가 모두 빈칸인지 확인..
        for d in range(4):
            ny, nx = cy + dy[d], cx + dx[d]
            # 이웃한 곳이 빈칸이 아니면 ,
            if nx>=0 and nx<M and ny>=0 and ny<N and m[ny][nx] != -1:
                impossible = False
                # 익은 토마토가 근처라면, 
                if m[ny][nx] == 1:
                    changed.append((cy, cx))        # 변화가 생겼고,
                    unrp.remove((cy,cx))  # 안익은 집합에서 지워줌..
                    break                 # 다음 안익은애 검사.
        if impossible and not unrp:
            changed = False
            break
    # 바뀐애들 익은걸로 update
    for y, x in changed:
        m[y][x] = 1
if changed:
    print(days)
else:
    print(-1)

6 4
1 -1 0 0 0 0
0 -1 0 0 0 0
0 0 0 0 -1 0
0 0 0 0 -1 1
안익은 애들: {(1, 2), (3, 2), (1, 3), (3, 3), (3, 0), (2, 2), (3, 1), (1, 4), (2, 1), (1, 5), (2, 0), (2, 5), (0, 5), (2, 3), (0, 4), (1, 0), (0, 2), (0, 3)}
6


시간초과ㅠㅠ

어떻게 해야 효율적으로 할까...
1. 안익은 애로 돌지 말고.. 가장자리에 있는 익은애들부터 시작해서 BFS

In [21]:
# import sys
# input = sys.stdin.readline
from collections import deque
dx = 0, 0, 1, -1
dy = 1, -1, 0, 0
M, N = map(int, input().split())
m = []

# 익은 애들 정보 저장하기..
q = deque()
n_unrp = 0
for j in range(N):
    r = list(map(int, input().split()))
    m.append(r)
    for i in range(M):
        if r[i]==1:
            q.append((j, i, 0))
        elif r[i] == 0:
            n_unrp += 1

# 안익은애가 없으면            
if not n_unrp:
    print(0)
    exit()

while q:
    y, x, t = q.popleft()
    for d in range(4):
        ny, nx = y + dy[d], x + dx[d]
        if nx>=0 and nx<M and ny>=0 and ny<N and m[ny][nx]==0:
            q.append((ny,nx,t+1))
            m[ny][nx] = 1
            n_unrp -= 1
if not n_unrp:
    print(t)
else:
    print(-1)


6 4
1 -1 0 0 0 0
0 -1 0 0 0 0
0 0 0 0 -1 0
0 0 0 0 -1 1
6


3036ms 한 두 배 느림.. 

In [ ]:
# 백준 다른사람 빠른 BFS
import sys
from collections import deque
input = sys.stdin.readline
def bfs(q):
    z = 0

    while q:
        x, y, z = q.popleft()
        # 조건을 이렇게 해주는게 더 빠르네..메..모..
        if x > 0 and t[x - 1][y] == 0:
            t[x - 1][y] = z + 1
            q.append((x - 1, y, z + 1))
        if x < n - 1 and t[x + 1][y] == 0:
            t[x + 1][y] = z + 1
            q.append((x + 1, y, z + 1))
        if y > 0 and t[x][y - 1] == 0:
            t[x][y - 1] = z + 1
            q.append((x, y - 1, z + 1))
        if y < m - 1 and t[x][y + 1] == 0:
            t[x][y + 1] = z + 1
            q.append((x, y + 1, z + 1))
    return z


m, n = map(int, input().split())
t = [[*map(int, input().split())] for _ in range(n)]
q = deque()

for i in range(n):
    for j in range(m):
        if t[i][j] >= 1:
            q.append((i, j, t[i][j]))

d = bfs(q)

for i in range(n):
    for j in range(m):
        if t[i][j] == 0:
            d = 0
            break

print(d - 1)


6 4  
0 0 0 0 0 0  
0 0 0 0 0 0  
0 0 0 0 0 0  
0 0 0 0 0 1  

---
8  
  
---  
6 4  
1 -1 0 0 0 0  
0 -1 0 0 0 0  
0 0 0 0 -1 0  
0 0 0 0 -1 1  

---
6

## 3. 2146 다리 만들기

#### 문제
여러 섬으로 이루어진 나라가 있다. 이 나라의 대통령은, 섬들을 잇는 다리를 만들겠다는 공약으로 인기몰이를 해 당선될 수 있었다.  
하지만 막상 대통령에 취임하자, 다리를 놓는다는 것이 아깝다는 생각을 하게 되었다.  
그래서 그는, 생색내는 식으로 한 섬과 다른 섬을 잇는 다리 하나만을 만들기로 하였고, 그 또한 다리를 가장 짧게 하여 돈을 아끼려 하였다.  
  
이 나라는 N×N크기의 이차원 평면상에 존재한다. 이 나라는 여러 섬으로 이루어져 있으며, 섬이란 동서남북으로 육지가 붙어있는 덩어리를 말한다.  
다음은 세 개의 섬으로 이루어진 나라의 지도이다.  
  
<img src="https://www.acmicpc.net/JudgeOnline/upload/201008/bri.PNG">
  
위의 그림에서 색이 있는 부분이 육지이고, 색이 없는 부분이 바다이다. 이 바다에 가장 짧은 다리를 놓아 두 대륙을 연결하고자 한다.  
가장 짧은 다리란, 다리가 격자에서 차지하는 칸의 수가 가장 작은 다리를 말한다. 다음 그림에서 두 대륙을 연결하는 다리를 볼 수 있다.  
  
<img src="https://www.acmicpc.net/JudgeOnline/upload/201008/b2.PNG">  
  
물론 위의 방법 외에도 다리를 놓는 방법이 여러 가지 있으나, 위의 경우가 놓는 다리의 길이가 3으로 가장 짧다.  
(물론 길이가 3인 다른 다리를 놓을 수 있는 방법도 몇 가지 있다.)  
  
지도가 주어질 때, 가장 짧은 다리 하나를 놓아 두 대륙을 연결하는 방법을 찾으시오.  
  
#### 입력
첫 줄에는 지도의 크기 N(100이하의 자연수)가 주어진다.  
그 다음 N줄에는 N개의 숫자가 빈칸을 사이에 두고 주어지며, 0은 바다, 1은 육지를 나타낸다.  
항상 두 개 이상의 섬이 있는 데이터만 입력으로 주어진다.  
  
#### 출력
첫째 줄에 가장 짧은 다리의 길이를 출력한다.  

### ------------------------------------------------------------------------------

일단 각각의 섬을 구분해야할듯?? 그리고 큐에 넣을때 (y, x, 섬번호, 길이) 이렇게 저장??   
바다에 붙어있는 애들 좌표만 각각의 큐에 저장. 다른 섬까지 다리를 이어보고 다리 길이 최소값 저장해서 그 이상 되면 그 섬에서는 탐색 종료..  

In [12]:
# import sys
# input = sys.stdin.readline
from copy import deepcopy
from collections import deque
N = int(input())
m = [list(map(int, input().split())) for _ in range(N)]
v = [[False]*N for _ in range(N)]
i_num = 1
edges = []
for j in range(N):
    for i in range(N):
        if not v[j][i] and m[j][i]:
            v[j][i] = True
            m[j][i] = i_num
            q = deque()
            edge = set()
            q.append((j,i))
            while q:
                y, x = q.popleft()
                if x>0 and not v[y][x-1]:
                    if m[y][x-1] == 1:
                        q.append((y, x-1))
                        v[y][x-1] = True
                        m[y][x-1] = i_num
                    if m[y][x-1] == 0:
                        edge.add((y,x))
                if x<N-1 and not v[y][x+1]:
                    if m[y][x+1] == 1:
                        q.append((y, x+1))
                        v[y][x+1] = True
                        m[y][x+1] = i_num
                    if m[y][x+1] == 0:
                        edge.add((y,x))
                if y>0 and not v[y-1][x]:
                    if m[y-1][x] == 1:
                        q.append((y-1, x))
                        v[y-1][x] = True
                        m[y-1][x] = i_num
                    if m[y-1][x] == 0:
                        edge.add((y,x))
                if y<N-1 and not v[y+1][x]:
                    if m[y+1][x] == 1:
                        q.append((y+1, x))
                        v[y+1][x] = True
                        m[y+1][x] = i_num
                    if m[y+1][x] == 0:
                        edge.add((y,x))
            edges.append(edge)
            i_num += 1
min_d = float('inf')
for i in range(1, len(edges)):
    v_temp = deepcopy(v)
    q = deque(edges[i-1])
    l = 0    
    while l<min_d:
        for _ in range(len(q)):
            y, x = q.popleft()
            if x>0:
                if m[y][x-1] == 0:
                    if not v[y][x-1]:
                        q.append((y, x-1))
                        v[y][x-1] = True
                elif m[y][x-1] != i:
                    min_d = l
                    break
            if x<N-1:
                if m[y][x+1] == 0:
                    if not v[y][x+1]:
                        q.append((y, x+1))
                        v[y][x+1] = True
                elif m[y][x+1] != i:
                    min_d = l
                    break
            if y>0:
                if m[y-1][x] == 0:
                    if not v[y-1][x]:
                        q.append((y-1, x))
                        v[y-1][x] = True
                elif m[y-1][x] != i:
                    min_d = l
                    break
            if y<N-1:
                if m[y+1][x] == 0:
                    if not v[y+1][x]:
                        q.append((y+1, x))
                        v[y+1][x] = True
                elif m[y+1][x] != i:
                    min_d = l
                    break                               
        l += 1
print(min_d)

5
1 1 0 0 1 
1 0 0 0 1
0 0 0 0 1
0 0 0 0 0
0 0 1 0 0
2


54%에서 시간초과.. ㅜㅜ deepcopy부분?? 섬번호 체크때문? 섬마다 돌아서?? ㅜㅜ  
각 edge 전체 한번에 돌아야하나.. 바다 탐색 큐에 어느섬에서 출발했는지도 저장해서??  

10   
1 1 1 0 0 0 0 1 1 1  
1 1 1 1 0 0 0 0 1 1  
1 0 1 1 0 0 0 0 1 1  
0 0 1 1 1 0 0 0 0 1  
0 0 0 1 0 0 0 0 0 1  
0 0 0 0 0 0 0 0 0 1  
0 0 0 0 0 0 0 0 0 0  
0 0 0 0 1 1 0 0 0 0  
0 0 0 0 1 1 1 0 0 0  
0 0 0 0 0 0 0 0 0 0  

---
3